pgmpy/metrics/PermutationTest.py

```diff
@@ -0,0 +1,160 @@
+from typing import Callable
+
+import numpy as np
+import pandas as pd
+from tqdm import tqdm
+
+from pgmpy import config
+from pgmpy.base import DAG
+from pgmpy.estimators.CITests import get_callable_ci_test
+from pgmpy.global_vars import logger
+from pgmpy.metrics import implied_cis
+from pgmpy.models import DiscreteBayesianNetwork
+
+
+def lmc_violations(model, ci_test_func: Callable, data, significance_level=0.05):
+    """
+    Counts LMC (Local Markov Chain) violations for a given DAG. For all ordered
+    triples of the form (node, non-descendant, parents) CI tests are performed
+    and based on significance level the violations are recorded.
+
+    This is based on https://ojs.aaai.org/index.php/AAAI/article/view/34881/37036
+
+    Parameters
+    ----------
+    model: A DAG for which we need to count the violations
+
+    ci_test_func: function
+        The function for statistical test. Can be either any of the tests in
+        pgmpy.estimators.CITests or any custom function of the same form.
+
+    data: pd.Dataframe
+        Dataset to use for testing
+
+    significance_level: p-value boundary
+        Used to check CI based on the p-value recieved from CI tests.
+
+    Returns
+    --------
+    violations: integer
+        The number of LMC violations found in the model.
+    """
+    violations = 0
+    nodes = list(model.nodes())
+    for node in nodes:
+        parents = list(model.get_parents(node))
+        non_descendants = model._get_non_descendants(node)
+        test_nodes = [(nd for nd in non_descendants if nd not in parents)]
+        for nd in test_nodes:
+            try:
+                p_value = ci_test_func(node, nd, parents, data, significance_level)
+                if not p_value:
+                    violations += 1
+            except Exception as e:
+                logger.debug(f"CI test failed for {node} ⊥ {nd} | {parents}: {e}")
+                continue
+
+    return violations
+
+
+def d_separated_triples(model, ci_test_func: Callable, data, significance_level=0.05):
+
+    cis_df = implied_cis(model, data, ci_test_func)
+    valid_cis = cis_df[cis_df["p-value"] >= significance_level]
+    dsep_triples = set()
+    for _, row in valid_cis.iterrows():
+        dsep_triples.add((row["u"], row["v"], frozenset(row["cond_vars"])))
+
+    return dsep_triples
+
+
+def permuted_graph(model, perm_mapping):
+    """
+    Creates node permutation over original model (DAG) while maintaining
+    structural/CI integrity of the model.
+
+    """
+    new_edges = []
+    for edge in model.edges():
+        new_source = perm_mapping[edge[0]]
+        new_target = perm_mapping[edge[1]]
+        new_edges.append((new_source, new_target))
+
+    permuted_model = DAG()
+    permuted_model.add_edges_from(new_edges)
+    return permuted_model
```

Comment:
We don't really need to create the permuted graph. The list of implied CIs remains the same in all the permuted graphs, just the variable name changes. 

For example, for a given graph `X -> Y -> Z` we have 1 implied CI: `X _|_ Z | Y`. If we consider a permutation mapping of `{X: Y, Y: X}`, the new graph would be: `Y -> X -> Z`, and our implied CI would be `Y _|_ Z | X`. So, instead of computing the permuted graph and then computing the implied CIs, we can apply the mapping to the implied CIs directly and it would give us the implied CIs of the permuted graph. This will be computationally much faster as well.